IfBuildProperty on classes with ServerRequestFilter causes build to fail

### Describe the bug

I have a ServerRequestFilter, which I only need to be active sometimes.
I therefore wanted to deactivate it by default with `@IfBuildProperty`.

```
@ApplicationScoped
@IfBuildProperty(name = "notexistingproperty", stringValue = "true")
public class MagicFilter {

    @ServerRequestFilter(priority = 0, preMatching = true)
    public void getFilter(HttpServerRequest request) {
        //TODO: do something..
    }
}
```

However, this results in the exception from below.

### Expected behavior

I can remove this bean with IfBuildProperty.

### Actual behavior

```
C:\workspaces\aviator\filter-ifproperty>mvn clean install
[INFO] Scanning for projects...
[INFO] 
[INFO] ---------------------< org.acme:filter-ifproperty >---------------------
[INFO] Building filter-ifproperty 1.0.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO] 
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ filter-ifproperty ---
[INFO] Deleting C:\workspaces\aviator\filter-ifproperty\target
[INFO] 
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ filter-ifproperty ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 2 resources
[INFO] 
[INFO] --- quarkus-maven-plugin:2.13.4.Final:generate-code (default) @ filter-ifproperty ---
[INFO] 
[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ filter-ifproperty ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 2 source files to C:\workspaces\aviator\filter-ifproperty\target\classes
[INFO] 
[INFO] --- quarkus-maven-plugin:2.13.4.Final:generate-code-tests (default) @ filter-ifproperty ---
[INFO] 
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ filter-ifproperty ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory C:\workspaces\aviator\filter-ifproperty\src\test\resources
[INFO]
[INFO] --- maven-compiler-plugin:3.8.1:testCompile (default-testCompile) @ filter-ifproperty ---
[INFO] No sources to compile
[INFO]
[INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ filter-ifproperty ---
[INFO] No tests to run.
[INFO]
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ filter-ifproperty ---
[INFO] Building jar: C:\workspaces\aviator\filter-ifproperty\target\filter-ifproperty-1.0.0-SNAPSHOT.jar
[INFO] 
[INFO] --- quarkus-maven-plugin:2.13.4.Final:build (default) @ filter-ifproperty ---
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  7.388 s
[INFO] Finished at: 2022-11-08T07:02:41+01:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal io.quarkus.platform:quarkus-maven-plugin:2.13.4.Final:build (default) on project filter-ifproperty: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR]         [error]: Build step io.quarkus.arc.deployment.ArcProcessor#validate threw an exception: javax.enterprise.inject.spi.DeploymentException: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type org.acme.MagicFilter and qualifiers [
@Default]
[ERROR]         - java member: org.acme.MagicFilter$GeneratedServerRequestFilter$getFilter():arg0
[ERROR]         - declared on CLASS bean [types=[javax.ws.rs.container.ContainerRequestFilter, org.acme.MagicFilter$GeneratedServerRequestFilter$getFilter, java.lang.Object], qualifiers=[@Default, @Any], target=org.acme.MagicFilter$GeneratedServerRequestFilter$getFilter]       
[ERROR]         at io.quarkus.arc.processor.BeanDeployment.processErrors(BeanDeployment.java:1217)
[ERROR]         at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:287)
[ERROR]         at io.quarkus.arc.processor.BeanProcessor.initialize(BeanProcessor.java:143)
[ERROR]         at io.quarkus.arc.deployment.ArcProcessor.validate(ArcProcessor.java:517)
[ERROR]         at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104)
[ERROR]         at java.base/java.lang.reflect.Method.invoke(Method.java:578)
[ERROR]         at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:909)
[ERROR]         at io.quarkus.builder.BuildContext.run(BuildContext.java:281)
[ERROR]         at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2449)
[ERROR]         at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1478)
[ERROR]         at java.base/java.lang.Thread.run(Thread.java:1589)
[ERROR]         at org.jboss.threads.JBossThread.run(JBossThread.java:501)
[ERROR] Caused by: javax.enterprise.inject.UnsatisfiedResolutionException: Unsatisfied dependency for type org.acme.MagicFilter and qualifiers [@Default]
[ERROR]         - java member: org.acme.MagicFilter$GeneratedServerRequestFilter$getFilter():arg0
[ERROR]         - declared on CLASS bean [types=[javax.ws.rs.container.ContainerRequestFilter, org.acme.MagicFilter$GeneratedServerRequestFilter$getFilter, java.lang.Object], qualifiers=[@Default, @Any], target=org.acme.MagicFilter$GeneratedServerRequestFilter$getFilter]       
[ERROR]         at io.quarkus.arc.processor.Beans.resolveInjectionPoint(Beans.java:440)
[ERROR]         at io.quarkus.arc.processor.BeanInfo.init(BeanInfo.java:539)
[ERROR]         at io.quarkus.arc.processor.BeanDeployment.init(BeanDeployment.java:275)
[ERROR]         ... 11 more
[ERROR] -> [Help 1]
[ERROR]
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR]
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException

````

### How to Reproduce?

Download the reproducer: 
[filter-ifproperty.zip](https://github.com/quarkusio/quarkus/files/9958123/filter-ifproperty.zip)

mvn clean install
exception from above happens

When you now remove the IfBuildProperty annotation from the MagicFIlter class, and build the project again, it works.

### Output of `uname -a` or `ver`

Microsoft Windows [Version 10.0.19044.2130]

### Output of `java -version`

openjdk 19 2022-09-20 OpenJDK Runtime Environment Temurin-19+36 (build 19+36) OpenJDK 64-Bit Server VM Temurin-19+36 (build 19+36, mixed mode, sharing)

### GraalVM version (if different from Java)

_No response_

### Quarkus version or git rev

2.13.4.Final

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.8.4 (9b656c72d54e5bacbed989b64718c159fe39b537) Maven home: C:\eclipse\tools\java\maven Java version: 19, vendor: Eclipse Adoptium, runtime: C:\eclipse\tools\java\19 Default locale: de_DE, platform encoding: UTF-8 OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"

### Additional information

_No response_

--- a/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/BuildTimeEnabledProcessor.java
+++ b/extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/BuildTimeEnabledProcessor.java
@@ -52,6 +52,9 @@ public class BuildTimeEnabledProcessor {
     private static final DotName UNLESS_BUILD_PROPERTY_CONTAINER = DotName
             .createSimple(UnlessBuildProperty.List.class.getName());
 
+    public static final Set<DotName> BUILD_TIME_ENABLED_BEAN_ANNOTATIONS = Set.of(IF_BUILD_PROFILE, UNLESS_BUILD_PROFILE,
+            IF_BUILD_PROPERTY, IF_BUILD_PROPERTY_CONTAINER, UNLESS_BUILD_PROPERTY, UNLESS_BUILD_PROPERTY_CONTAINER);
+
     @BuildStep
     void ifBuildProfile(CombinedIndexBuildItem index, BuildProducer<BuildTimeConditionBuildItem> producer) {
         Collection<AnnotationInstance> annotationInstances = index.getIndex().getAnnotations(IF_BUILD_PROFILE);

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/ResteasyReactiveScanningProcessor.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/ResteasyReactiveScanningProcessor.java
@@ -47,6 +47,7 @@ import org.jboss.resteasy.reactive.server.processor.scanning.ResteasyReactiveFea
 import io.quarkus.arc.ArcUndeclaredThrowableException;
 import io.quarkus.arc.Unremovable;
 import io.quarkus.arc.deployment.AdditionalBeanBuildItem;
+import io.quarkus.arc.deployment.BuildTimeEnabledProcessor;
 import io.quarkus.arc.deployment.GeneratedBeanBuildItem;
 import io.quarkus.arc.deployment.GeneratedBeanGizmoAdaptor;
 import io.quarkus.arc.processor.DotNames;
@@ -310,7 +311,16 @@ public class ResteasyReactiveScanningProcessor {
         }
 
         List<FilterGeneration.GeneratedFilter> generatedFilters = FilterGeneration.generate(index,
-                Set.of(HTTP_SERVER_REQUEST, HTTP_SERVER_RESPONSE, ROUTING_CONTEXT), Set.of(Unremovable.class.getName()));
+                Set.of(HTTP_SERVER_REQUEST, HTTP_SERVER_RESPONSE, ROUTING_CONTEXT), Set.of(Unremovable.class.getName()),
+                (methodInfo -> {
+                    List<AnnotationInstance> classAnnotations = methodInfo.declaringClass().declaredAnnotations();
+                    for (AnnotationInstance classAnnotation : classAnnotations) {
+                        if (BuildTimeEnabledProcessor.BUILD_TIME_ENABLED_BEAN_ANNOTATIONS.contains(classAnnotation.name())) {
+                            return true;
+                        }
+                    }
+                    return false;
+                }));
         for (var generated : generatedFilters) {
             for (var i : generated.getGeneratedClasses()) {
                 generatedBean.produce(new GeneratedBeanBuildItem(i.getName(), i.getData()));

--- a/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/customproviders/ConditionalBeanFiltersTest.java
+++ b/extensions/resteasy-reactive/quarkus-resteasy-reactive/deployment/src/test/java/io/quarkus/resteasy/reactive/server/test/customproviders/ConditionalBeanFiltersTest.java
@@ -1,0 +1,149 @@
+package io.quarkus.resteasy.reactive.server.test.customproviders;
+
+import static io.restassured.RestAssured.*;
+import static org.assertj.core.api.Assertions.*;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+
+import java.util.List;
+import java.util.Optional;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import javax.ws.rs.GET;
+import javax.ws.rs.Path;
+import javax.ws.rs.Priorities;
+import javax.ws.rs.container.ContainerRequestContext;
+import javax.ws.rs.container.ContainerResponseContext;
+import javax.ws.rs.core.HttpHeaders;
+import javax.ws.rs.core.Response;
+
+import org.hamcrest.Matchers;
+import org.jboss.resteasy.reactive.server.ServerRequestFilter;
+import org.jboss.resteasy.reactive.server.ServerResponseFilter;
+import org.jboss.shrinkwrap.api.ShrinkWrap;
+import org.jboss.shrinkwrap.api.spec.JavaArchive;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import io.quarkus.arc.profile.IfBuildProfile;
+import io.quarkus.arc.properties.IfBuildProperty;
+import io.quarkus.test.QuarkusUnitTest;
+import io.restassured.http.Header;
+import io.smallrye.mutiny.Uni;
+
+public class ConditionalBeanFiltersTest {
+
+    @RegisterExtension
+    static QuarkusUnitTest test = new QuarkusUnitTest()
+            .setArchiveProducer(new Supplier<>() {
+                @Override
+                public JavaArchive get() {
+                    return ShrinkWrap.create(JavaArchive.class)
+                            .addClasses(WontBeEnabledFilter.class, WillBeEnabledFilter.class, AlwaysEnabledFilter.class,
+                                    TestResource.class);
+                }
+            });
+
+    @Test
+    public void testExpectedFilters() {
+        List<String> responseFiltersValues = get("/test/filters")
+                .then().statusCode(200)
+                .body(Matchers.is("void-on,response-on,uni-on,always"))
+                .extract()
+                .headers()
+                .getList("response-filters")
+                .stream()
+                .map(Header::getValue)
+                .collect(Collectors.toList());
+        assertThat(responseFiltersValues).containsOnly("always", "void-on", "uni-on");
+    }
+
+    @Path("test")
+    public static class TestResource {
+
+        @Path("filters")
+        @GET
+        public String filters(HttpHeaders headers) {
+            return String.join(",", headers.getRequestHeader("request-filters"));
+        }
+    }
+
+    @IfBuildProperty(name = "notexistingproperty", stringValue = "true")
+    public static class WontBeEnabledFilter {
+
+        @ServerRequestFilter(priority = Priorities.USER + 1)
+        public void voidRequestFilter(ContainerRequestContext requestContext) {
+            requestContext.getHeaders().add("request-filters", "void-off");
+        }
+
+        @ServerRequestFilter(priority = Priorities.USER + 2)
+        public Response responseTypeRequestFilter(ContainerRequestContext requestContext) {
+            requestContext.getHeaders().add("request-filters", "response-off");
+            return null;
+        }
+
+        @ServerRequestFilter(priority = Priorities.USER + 3)
+        public Uni<Void> uniRequestFilter(ContainerRequestContext requestContext) {
+            requestContext.getHeaders().add("request-filters", "uni-off");
+            return Uni.createFrom().nullItem();
+        }
+
+        // if any of the following were to be executed, they would fail, thrown an exception and result in an HTTP 500
+
+        @ServerResponseFilter
+        public void voidResponseFilter(ContainerResponseContext ctx) {
+            assertFalse(true);
+        }
+
+        @ServerResponseFilter
+        public Uni<Void> uniResponseFilter(ContainerResponseContext ctx) {
+            assertFalse(true);
+            return Uni.createFrom().nullItem();
+        }
+    }
+
+    @IfBuildProfile("test")
+    public static class WillBeEnabledFilter {
+
+        @ServerRequestFilter(priority = Priorities.USER + 4)
+        public void voidRequestFilter(ContainerRequestContext requestContext) {
+            requestContext.getHeaders().add("request-filters", "void-on");
+        }
+
+        @ServerRequestFilter(priority = Priorities.USER + 5)
+        public Optional<Response> responseTypeRequestFilter(ContainerRequestContext requestContext) {
+            requestContext.getHeaders().add("request-filters", "response-on");
+            return Optional.empty();
+        }
+
+        @ServerRequestFilter(priority = Priorities.USER + 6)
+        public Uni<Response> uniRequestFilter(ContainerRequestContext requestContext) {
+            requestContext.getHeaders().add("request-filters", "uni-on");
+            return Uni.createFrom().nullItem();
+        }
+
+        @ServerResponseFilter(priority = Priorities.USER + 4)
+        public void voidResponseFilter(ContainerResponseContext ctx) {
+            ctx.getHeaders().add("response-filters", "void-on");
+        }
+
+        @ServerResponseFilter(priority = Priorities.USER + 6)
+        public Uni<Void> uniResponseFilter(ContainerResponseContext ctx) {
+            ctx.getHeaders().add("response-filters", "uni-on");
+            return Uni.createFrom().nullItem();
+        }
+    }
+
+    public static class AlwaysEnabledFilter {
+
+        @ServerRequestFilter(priority = Priorities.USER + 100)
+        public void alwaysRequestFilter(ContainerRequestContext requestContext) {
+            requestContext.getHeaders().add("request-filters", "always");
+        }
+
+        @ServerResponseFilter(priority = Priorities.USER + 100)
+        public void voidResponseFilter(ContainerResponseContext ctx) {
+            ctx.getHeaders().add("response-filters", "always");
+        }
+    }
+}

--- a/independent-projects/resteasy-reactive/server/processor/src/main/java/org/jboss/resteasy/reactive/server/processor/generation/filters/CustomFilterGenerator.java
+++ b/independent-projects/resteasy-reactive/server/processor/src/main/java/org/jboss/resteasy/reactive/server/processor/generation/filters/CustomFilterGenerator.java
@@ -17,9 +17,11 @@ import java.lang.reflect.Modifier;
 import java.util.List;
 import java.util.Optional;
 import java.util.Set;
+import java.util.function.Predicate;
 
 import javax.enterprise.context.ApplicationScoped;
 import javax.enterprise.context.RequestScoped;
+import javax.enterprise.inject.Instance;
 import javax.inject.Inject;
 import javax.inject.Singleton;
 import javax.ws.rs.container.ContainerRequestContext;
@@ -56,6 +58,7 @@ import org.jboss.resteasy.reactive.server.spi.ServerRequestContext;
 
 import io.quarkus.gizmo.AssignableResultHandle;
 import io.quarkus.gizmo.BranchResult;
+import io.quarkus.gizmo.BytecodeCreator;
 import io.quarkus.gizmo.ClassCreator;
 import io.quarkus.gizmo.ClassOutput;
 import io.quarkus.gizmo.FieldDescriptor;
@@ -75,21 +78,25 @@ final class CustomFilterGenerator {
 
     private final Set<DotName> unwrappableTypes;
     private final Set<String> additionalBeanAnnotations;
+    private final Predicate<MethodInfo> isOptionalFilter;
 
-    CustomFilterGenerator(Set<DotName> unwrappableTypes, Set<String> additionalBeanAnnotations) {
+    CustomFilterGenerator(Set<DotName> unwrappableTypes, Set<String> additionalBeanAnnotations,
+            Predicate<MethodInfo> isOptionalFilter) {
         this.unwrappableTypes = unwrappableTypes;
         this.additionalBeanAnnotations = additionalBeanAnnotations;
+        this.isOptionalFilter = isOptionalFilter;
     }
 
     String generateContainerRequestFilter(MethodInfo targetMethod, ClassOutput classOutput) {
         checkModifiers(targetMethod, ResteasyReactiveServerDotNames.SERVER_REQUEST_FILTER);
         if (KotlinUtils.isSuspendMethod(targetMethod)) {
-            return generateRequestFilterForSuspendedMethod(targetMethod, classOutput);
+            return generateRequestFilterForSuspendedMethod(targetMethod, classOutput, isOptionalFilter.test(targetMethod));
         }
-        return generateStandardRequestFilter(targetMethod, classOutput);
+        return generateStandardRequestFilter(targetMethod, classOutput, isOptionalFilter.test(targetMethod));
     }
 
-    private String generateRequestFilterForSuspendedMethod(MethodInfo targetMethod, ClassOutput classOutput) {
+    private String generateRequestFilterForSuspendedMethod(MethodInfo targetMethod, ClassOutput classOutput,
+            boolean checkForOptionalBean) {
         DotName returnDotName = determineReturnDotNameOfSuspendMethod(targetMethod);
         ReturnType returnType;
         if (returnDotName.equals(VOID)) {
@@ -112,14 +119,21 @@ final class CustomFilterGenerator {
                 .superClass(ABSTRACT_SUSPENDED_REQ_FILTER)
                 .build()) {
             FieldDescriptor delegateField = generateConstructorAndDelegateField(cc, declaringClass,
-                    ABSTRACT_SUSPENDED_REQ_FILTER, additionalBeanAnnotations);
+                    ABSTRACT_SUSPENDED_REQ_FILTER, additionalBeanAnnotations, checkForOptionalBean);
 
             // generate the implementation of the 'doFilter' method
             MethodCreator doFilterMethod = cc.getMethodCreator("doFilter", Object.class.getName(),
                     ResteasyReactiveContainerRequestContext.class.getName(), ResteasyReactiveDotNames.CONTINUATION.toString());
+            ResultHandle delegate = doFilterMethod.readInstanceField(delegateField, doFilterMethod.getThis());
+
+            if (checkForOptionalBean) {
+                // if the delegate is null (because there was no bean of the target type, just return),
+                doFilterMethod.ifNull(delegate).trueBranch().returnNull();
+            }
+
             // call the target method
             ResultHandle resultHandle = doFilterMethod.invokeVirtualMethod(targetMethod,
-                    doFilterMethod.readInstanceField(delegateField, doFilterMethod.getThis()),
+                    delegate,
                     getRequestFilterResultHandles(targetMethod, declaringClassName, doFilterMethod, 2,
                             getRRReqCtxHandle(doFilterMethod, getRRContainerReqCtxHandle(doFilterMethod, 0))));
             doFilterMethod.returnValue(resultHandle);
@@ -151,7 +165,7 @@ final class CustomFilterGenerator {
                                     Uni.class, ResteasyReactiveContainerRequestContext.class),
                             handleResultMethod.getMethodParam(1), getRRContainerReqCtxHandle(handleResultMethod, 0));
 
-            handleResultMethod.returnValue(null);
+            handleResultMethod.returnNull();
         }
 
         return generatedClassName;
@@ -186,7 +200,8 @@ final class CustomFilterGenerator {
      *
      * </pre>
      */
-    private String generateStandardRequestFilter(MethodInfo targetMethod, ClassOutput classOutput) {
+    private String generateStandardRequestFilter(MethodInfo targetMethod, ClassOutput classOutput,
+            boolean checkForOptionalBean) {
         ReturnType returnType = determineRequestFilterReturnType(targetMethod);
         String generatedClassName = getGeneratedClassName(targetMethod, ResteasyReactiveServerDotNames.SERVER_REQUEST_FILTER);
         ClassInfo declaringClass = targetMethod.declaringClass();
@@ -196,7 +211,7 @@ final class CustomFilterGenerator {
                 .interfaces(determineRequestInterfaceType(returnType))
                 .build()) {
             FieldDescriptor delegateField = generateConstructorAndDelegateField(cc, declaringClass,
-                    Object.class.getName(), additionalBeanAnnotations);
+                    Object.class.getName(), additionalBeanAnnotations, checkForOptionalBean);
 
             if (returnType == ReturnType.VOID
                     || returnType == ReturnType.OPTIONAL_RESPONSE
@@ -205,10 +220,23 @@ final class CustomFilterGenerator {
                     || returnType == ReturnType.REST_RESPONSE) {
                 // generate the implementation of the filter method
                 MethodCreator filterMethod = cc.getMethodCreator("filter", void.class, ContainerRequestContext.class);
+                ResultHandle delegate = filterMethod.readInstanceField(delegateField, filterMethod.getThis());
+
+                if (checkForOptionalBean) {
+                    // if the delegate is null (because there was no bean of the target type, just return),
+                    BytecodeCreator delegateIsNull = filterMethod.ifNull(delegate).trueBranch();
+                    if (returnType == ReturnType.VOID) {
+                        delegateIsNull.returnVoid();
+                    } else {
+                        delegateIsNull.returnNull();
+                    }
+                }
+
                 ResultHandle rrContainerReqCtxHandle = getRRContainerReqCtxHandle(filterMethod, 0);
                 // call the target method
+
                 ResultHandle resultHandle = filterMethod.invokeVirtualMethod(targetMethod,
-                        filterMethod.readInstanceField(delegateField, filterMethod.getThis()),
+                        delegate,
                         getRequestFilterResultHandles(targetMethod, declaringClassName, filterMethod, 1,
                                 getRRReqCtxHandle(filterMethod, rrContainerReqCtxHandle)));
                 if (returnType == ReturnType.OPTIONAL_RESPONSE) {
@@ -241,6 +269,13 @@ final class CustomFilterGenerator {
                 // generate the implementation of the filter method
                 MethodCreator filterMethod = cc.getMethodCreator("filter", void.class,
                         ResteasyReactiveContainerRequestContext.class);
+
+                if (checkForOptionalBean) {
+                    // if the delegate is null (because there was no bean of the target type, just return),
+                    filterMethod.ifNull(filterMethod.readInstanceField(delegateField, filterMethod.getThis())).trueBranch()
+                            .returnNull();
+                }
+
                 // call the target method
                 ResultHandle rrContainerReqCtxHandle = getRRContainerReqCtxHandle(filterMethod, 0);
                 ResultHandle uniHandle = filterMethod.invokeVirtualMethod(targetMethod,
@@ -270,7 +305,7 @@ final class CustomFilterGenerator {
                                         void.class,
                                         Uni.class, ResteasyReactiveContainerRequestContext.class),
                                 uniHandle, rrContainerReqCtxHandle);
-                filterMethod.returnValue(null);
+                filterMethod.returnNull();
             } else {
                 throw new IllegalStateException("ReturnType: '" + returnType + "' is not supported, in method "
                         + targetMethod.declaringClass() + "." + targetMethod.name());
@@ -332,12 +367,13 @@ final class CustomFilterGenerator {
     String generateContainerResponseFilter(MethodInfo targetMethod, ClassOutput classOutput) {
         checkModifiers(targetMethod, ResteasyReactiveServerDotNames.SERVER_RESPONSE_FILTER);
         if (KotlinUtils.isSuspendMethod(targetMethod)) {
-            return generateResponseFilterForSuspendedMethod(targetMethod, classOutput);
+            return generateResponseFilterForSuspendedMethod(targetMethod, classOutput, isOptionalFilter.test(targetMethod));
         }
-        return generateStandardContainerResponseFilter(targetMethod, classOutput);
+        return generateStandardContainerResponseFilter(targetMethod, classOutput, isOptionalFilter.test(targetMethod));
     }
 
-    private String generateResponseFilterForSuspendedMethod(MethodInfo targetMethod, ClassOutput classOutput) {
+    private String generateResponseFilterForSuspendedMethod(MethodInfo targetMethod, ClassOutput classOutput,
+            boolean checkForOptionalBean) {
         DotName returnDotName = determineReturnDotNameOfSuspendMethod(targetMethod);
         if (!returnDotName.equals(VOID)) {
             throw new RuntimeException("Suspend method '" + targetMethod.name() + " of class '"
@@ -352,7 +388,7 @@ final class CustomFilterGenerator {
                 .superClass(ABSTRACT_SUSPENDED_RES_FILTER)
                 .build()) {
             FieldDescriptor delegateField = generateConstructorAndDelegateField(cc, declaringClass,
-                    ABSTRACT_SUSPENDED_RES_FILTER, additionalBeanAnnotations);
+                    ABSTRACT_SUSPENDED_RES_FILTER, additionalBeanAnnotations, checkForOptionalBean);
 
             // generate the implementation of the filter method
             MethodCreator doFilterMethod = cc.getMethodCreator("doFilter", Object.class.getName(),
@@ -401,7 +437,8 @@ final class CustomFilterGenerator {
      *
      * </pre>
      */
-    private String generateStandardContainerResponseFilter(MethodInfo targetMethod, ClassOutput classOutput) {
+    private String generateStandardContainerResponseFilter(MethodInfo targetMethod, ClassOutput classOutput,
+            boolean checkForOptionalBean) {
         ReturnType returnType = determineResponseFilterReturnType(targetMethod);
         String generatedClassName = getGeneratedClassName(targetMethod, ResteasyReactiveServerDotNames.SERVER_RESPONSE_FILTER);
         ClassInfo declaringClassInfo = targetMethod.declaringClass();
@@ -411,12 +448,17 @@ final class CustomFilterGenerator {
                 .interfaces(determineResponseInterfaceType(returnType))
                 .build()) {
             FieldDescriptor delegateField = generateConstructorAndDelegateField(cc, declaringClassInfo,
-                    Object.class.getName(), additionalBeanAnnotations);
+                    Object.class.getName(), additionalBeanAnnotations, checkForOptionalBean);
 
             if (returnType == ReturnType.VOID) {
                 // generate the implementation of the filter method
                 MethodCreator filterMethod = cc.getMethodCreator("filter", void.class, ContainerRequestContext.class,
                         ContainerResponseContext.class);
+
+                if (checkForOptionalBean) {
+                    filterMethod.ifNull(filterMethod.readInstanceField(delegateField, filterMethod.getThis())).trueBranch()
+                            .returnVoid();
+                }
 
                 ResultHandle rrContainerReqCtxHandle = getRRContainerReqCtxHandle(filterMethod, 0);
                 ResultHandle rrReqCtxHandle = getRRReqCtxHandle(filterMethod, rrContainerReqCtxHandle);
@@ -432,6 +474,11 @@ final class CustomFilterGenerator {
                 MethodCreator filterMethod = cc.getMethodCreator("filter", void.class,
                         ResteasyReactiveContainerRequestContext.class,
                         ContainerResponseContext.class);
+
+                if (checkForOptionalBean) {
+                    filterMethod.ifNull(filterMethod.readInstanceField(delegateField, filterMethod.getThis())).trueBranch()
+                            .returnNull();
+                }
 
                 ResultHandle rrContainerReqCtxHandle = getRRContainerReqCtxHandle(filterMethod, 0);
                 ResultHandle rrReqCtxHandle = getRRReqCtxHandle(filterMethod, rrContainerReqCtxHandle);
@@ -454,7 +501,8 @@ final class CustomFilterGenerator {
     }
 
     private FieldDescriptor generateConstructorAndDelegateField(ClassCreator cc, ClassInfo declaringClass,
-            String superClassName, Set<String> additionalBeanAnnotations) {
+            String superClassName, Set<String> additionalBeanAnnotations,
+            boolean checkForOptionalBean) {
 
         String declaringClassName = declaringClass.toString();
         ScopeInspectionResult scopeInspectionResult = inspectScope(declaringClass);
@@ -470,14 +518,35 @@ final class CustomFilterGenerator {
                 .setModifiers(Modifier.PRIVATE | Modifier.FINAL)
                 .getFieldDescriptor();
 
-        // generate a constructor that takes the target class as an argument - this class is a CDI bean so Arc will be able to inject into the generated class
-        MethodCreator ctor = cc.getMethodCreator("<init>", void.class, declaringClassName);
+        MethodCreator ctor;
+        if (checkForOptionalBean) {
+            // generate a constructor that takes the Instance<TargetClass> as an argument
+            ctor = cc.getMethodCreator("<init>", void.class, Instance.class).setSignature(
+                    String.format("(Ljavax/enterprise/inject/Instance<L%s;>;)V", declaringClassName.replace('.', '/')));
+        } else {
+            // generate a constructor that takes the target class as an argument - this class is a CDI bean so Arc will be able to inject into the generated class
+            ctor = cc.getMethodCreator("<init>", void.class, declaringClassName);
+        }
         ctor.setModifiers(Modifier.PUBLIC);
         ctor.addAnnotation(Inject.class);
         ctor.invokeSpecialMethod(MethodDescriptor.ofConstructor(superClassName), ctor.getThis());
         ResultHandle self = ctor.getThis();
-        ResultHandle delegate = ctor.getMethodParam(0);
-        ctor.writeInstanceField(delegateField, self, delegate);
+        if (checkForOptionalBean) {
+            ResultHandle instance = ctor.getMethodParam(0);
+            ResultHandle isResolvable = ctor
+                    .invokeInterfaceMethod(MethodDescriptor.ofMethod(Instance.class, "isResolvable", boolean.class), instance);
+            BranchResult isResolvableBranch = ctor.ifTrue(isResolvable);
+
+            BytecodeCreator isResolvableTrue = isResolvableBranch.trueBranch();
+            isResolvableTrue.writeInstanceField(delegateField, self, isResolvableTrue
+                    .invokeInterfaceMethod(MethodDescriptor.ofMethod(Instance.class, "get", Object.class), instance));
+
+            BytecodeCreator isResolvableFalse = isResolvableBranch.falseBranch();
+            isResolvableFalse.writeInstanceField(delegateField, self, isResolvableFalse.loadNull());
+        } else {
+            ctor.writeInstanceField(delegateField, self, ctor.getMethodParam(0));
+        }
+
         ctor.returnValue(null);
 
         if (scopeInspectionResult.needsProxy) {

--- a/independent-projects/resteasy-reactive/server/processor/src/main/java/org/jboss/resteasy/reactive/server/processor/generation/filters/FilterFeature.java
+++ b/independent-projects/resteasy-reactive/server/processor/src/main/java/org/jboss/resteasy/reactive/server/processor/generation/filters/FilterFeature.java
@@ -34,7 +34,7 @@ public class FilterFeature implements FeatureScanner {
     public FeatureScanResult integrate(IndexView application, ScannedApplication scannedApplication) {
         List<GeneratedClass> generatedClasses = new ArrayList<>();
         List<FilterGeneration.GeneratedFilter> result = FilterGeneration.generate(application, unwrappableTypes,
-                additionalBeanAnnotations);
+                additionalBeanAnnotations, (m) -> false);
         for (var i : result) {
             generatedClasses.addAll(i.getGeneratedClasses());
             if (i.isRequestFilter()) {

--- a/independent-projects/resteasy-reactive/server/processor/src/main/java/org/jboss/resteasy/reactive/server/processor/generation/filters/FilterGeneration.java
+++ b/independent-projects/resteasy-reactive/server/processor/src/main/java/org/jboss/resteasy/reactive/server/processor/generation/filters/FilterGeneration.java
@@ -7,6 +7,7 @@ import java.util.ArrayList;
 import java.util.HashSet;
 import java.util.List;
 import java.util.Set;
+import java.util.function.Predicate;
 
 import org.jboss.jandex.AnnotationInstance;
 import org.jboss.jandex.AnnotationTarget;
@@ -21,7 +22,7 @@ import org.jboss.resteasy.reactive.server.processor.util.GeneratedClassOutput;
 
 public class FilterGeneration {
     public static List<GeneratedFilter> generate(IndexView index, Set<DotName> unwrappableTypes,
-            Set<String> additionalBeanAnnotations) {
+            Set<String> additionalBeanAnnotations, Predicate<MethodInfo> isOptionalFilter) {
         List<GeneratedFilter> ret = new ArrayList<>();
         for (AnnotationInstance instance : index
                 .getAnnotations(SERVER_REQUEST_FILTER)) {
@@ -30,7 +31,7 @@ public class FilterGeneration {
             }
             MethodInfo methodInfo = instance.target().asMethod();
             GeneratedClassOutput output = new GeneratedClassOutput();
-            String generatedClassName = new CustomFilterGenerator(unwrappableTypes, additionalBeanAnnotations)
+            String generatedClassName = new CustomFilterGenerator(unwrappableTypes, additionalBeanAnnotations, isOptionalFilter)
                     .generateContainerRequestFilter(methodInfo, output);
             Integer priority = null;
             boolean preMatching = false;
@@ -88,7 +89,8 @@ public class FilterGeneration {
             Integer priority = null;
             Set<String> nameBindingNames = new HashSet<>();
             GeneratedClassOutput output = new GeneratedClassOutput();
-            String generatedClassName = new CustomFilterGenerator(unwrappableTypes, additionalBeanAnnotations)
+            String generatedClassName = new CustomFilterGenerator(unwrappableTypes, additionalBeanAnnotations,
+                    isOptionalFilter)
                     .generateContainerResponseFilter(methodInfo, output);
 
             AnnotationValue priorityValue = instance.value("priority");

